### PR TITLE
Update Fold + Env to allow for filters for instance status

### PR DESCRIPTION
**Pull request by oswynb**
Add a 'Filters' to the Env to supply ceilometer-common with a predicate to filter Instance metrics based on status.

### CODE REVIEW

**Comment by tranma** [comment]
I'm not sure how this is supposed to work. Say you have events:

```
                                                 query end somewhere here
                                                 v
InstanceActive | InstanceSuspended | InstanceResize 
       ^ 
       query start somewhere here
```

and `InstanceSuspended` is not "billable", so you just ignore the time delta between `start` and `InstanceSuspended`, but you don't change the accumulator, so you end up adding the delta between `start` and `InstanceResize` -- that wouldn't change the result at all, would it?

Can you put an example here so we can see?

**Comment by tranma** [comment]
Ah, I see. It's a bit convoluted, I can't think of a cleaner way on the top of my head. Regardless, could you add some quickcheck tests, for both instance flavors and non-flavor resources?

**Comment by tranma** [comment]
if we have quickcheck tests, we can formulate exactly what we're expecting here.

**Comment by oswynb** [comment]
Shall do

**Comment by tranma** [comment]
tests/SampleData has all the Arbitrary & friends instances

**Comment by olorin** [comment]
I think this needs a minor bump?

File: tests/SampleData.hs [hunk 1]
```diff
@@ -193,6 +207,20 @@ flavorTimedPDsResult :: [(PFValue PDInstanceFlavor, Word64)]
...
+        fullResult     = L.fold (foldInstanceFlavor (const True)) x
+        filteredResult = L.fold (foldInstanceFlavor testPred    ) x
+        fullKeys       = M.keys fullResult
+        filteredKeys   = M.keys filteredResult
+    in  subset filteredKeys fullKeys &&
+        all (\k -> isJust $ do
+                v  <- M.lookup k fullResult
+                v' <- M.lookup k filteredResult
+                return $ v' <= v) filteredKeys
+
```
Comment (tranma) [inline tests/SampleData.hs:229]: can you put a short comment describing this property -- something to the effect of "filtered results must be subset of ... and all filtered values are less than corresponding values in full results"

then please merge.

**Comment by tranma** [comment]
sign-off.

File: tests/Fold.hs [hunk 2]
```diff
@@ -85,7 +85,10 @@ suite = do
   -- "Gauge Pollster" resources
   describe "Folding points for POLLSTER resource: INSTANCE FLAVOR" $
     it "ok for example payload"
-      $ L.fold foldInstanceFlavor flavorTimedPDs `shouldBe` M.fromList flavorTimedPDsResult
+      $ L.fold (foldInstanceFlavor $ const True) flavorTimedPDs `shouldBe` M.fromList flavorTimedPDsResult
+
+  describe "Filtering instance statuses is sane" $
```
Comment (olorin) [inline tests/Fold.hs]: This use of "describe" is odd - normally, describe [is used to name the thing which is described by the tests](https://hspec.github.io/), not assert that the thing is sane.

Comment (tranma) [inline tests/Fold.hs]: describe "filtering instance statuses" $
  prop "is sane" ....

Comment (olorin) [inline tests/Fold.hs]: Yep.

**Pull request merged**